client/src/main/java/io/pravega/client/batch/impl/BatchClientImpl.java

```diff
@@ -0,0 +1,102 @@
+/**
+ * Copyright (c) 2017 Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.client.batch.impl;
+
+import com.google.common.annotations.Beta;
+import com.google.common.collect.Iterators;
+import io.pravega.client.batch.BatchClient;
+import io.pravega.client.batch.SegmentInfo;
+import io.pravega.client.batch.SegmentIterator;
+import io.pravega.client.batch.StreamInfo;
+import io.pravega.client.netty.impl.ConnectionFactory;
+import io.pravega.client.segment.impl.Segment;
+import io.pravega.client.segment.impl.SegmentInputStreamFactory;
+import io.pravega.client.segment.impl.SegmentInputStreamFactoryImpl;
+import io.pravega.client.segment.impl.SegmentMetadataClient;
+import io.pravega.client.segment.impl.SegmentMetadataClientFactory;
+import io.pravega.client.segment.impl.SegmentMetadataClientFactoryImpl;
+import io.pravega.client.stream.Serializer;
+import io.pravega.client.stream.Stream;
+import io.pravega.client.stream.impl.Controller;
+import io.pravega.client.stream.impl.StreamCut;
+import io.pravega.client.stream.impl.StreamImpl;
+import java.util.Date;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.SortedSet;
+import java.util.TreeSet;
+import lombok.Cleanup;
+import org.apache.commons.lang3.NotImplementedException;
+
+import static io.pravega.common.concurrent.FutureHelpers.getAndHandleExceptions;
+
+@Beta
+public class BatchClientImpl implements BatchClient {
+
+    private final Controller controller;
+    private final SegmentInputStreamFactory inputStreamFactory;
+    private final SegmentMetadataClientFactory segmentMetadataClientFactory;
+
+    public BatchClientImpl(Controller controller, ConnectionFactory connectionFactory) {
+        this.controller = controller;
+        inputStreamFactory = new SegmentInputStreamFactoryImpl(controller, connectionFactory);
+        segmentMetadataClientFactory = new SegmentMetadataClientFactoryImpl(controller, connectionFactory);
+    }
+
+    @Override
+    public StreamInfo getStreamInfo(Stream stream) {
+        // TODO: Implement this method.
+        // Name from stream
+        // Length refector from ReaderGroupImpl perhaps move to controller.
+        // Creation time needs an added api? or perhaps modify the getsegmentAtTime api
+        // create a controller.getStreamSealTime() which returns null if open
+        throw new NotImplementedException("getStreamInfo");
+    }
+
+    @Override
+    public Iterator<SegmentInfo> listSegments(Stream stream) {
+        return listSegments(stream, new Date(0L));
+    }
+
+    @Override
+    public Iterator<SegmentInfo> listSegments(Stream stream, Date from) {
```

Comment:
I'm worried that the granularity of the time argument in the `getSegmentsAtTime` is too coarse and won't be satisfactory. I think people will expect some form of time index for this, so I'm wondering whether we should have this call at all at this point.